# HG changeset patch
# User Torsten <ttl@justmail.de>
# Date 1437740608 -7200
#      Fri Jul 24 14:23:28 2015 +0200
# Branch stable
# Node ID 2b28f509b64537495687084c0c9c29f02c762cbb
# Parent  587357240e5e80e4c771c5089a97769231080b4c
Fix missing sorting of the workspace view directly after start-up

* workspace-view.cc(workspace_view): add sorting indicator, do not set the
  column by which the view is sorted here;
  (setModel) but here, after having set the model

diff --git a/libgui/src/workspace-view.cc b/libgui/src/workspace-view.cc
--- a/libgui/src/workspace-view.cc
+++ b/libgui/src/workspace-view.cc
@@ -87,28 +87,30 @@ workspace_view::workspace_view (QWidget 
 
   // Set the empty widget to have our layout.
   widget ()->setLayout (vbox_layout);
 
   // Initialize collapse/expand state of the workspace subcategories.
 
   QSettings *settings = resource_manager::get_settings ();
 
-  //enable sorting as previously configured
+  //enable sorting (setting column and order after model was set)
   view->setSortingEnabled (true);
-  view->sortByColumn (
+  // Initialize column order and width of the workspace
+  view->horizontalHeader ()->restoreState (
+    settings->value ("workspaceview/column_state").toByteArray ());
+  // Set header properties for sorting
+  view->horizontalHeader ()->setClickable (true);
+  view->horizontalHeader ()->setMovable (true);
+  view->horizontalHeader ()->setSortIndicator (
     settings->value ("workspaceview/sort_by_column",0).toInt (),
     static_cast<Qt::SortOrder>
     (settings->value ("workspaceview/sort_order", Qt::AscendingOrder).toUInt ())
   );
-  // Initialize column order and width of the workspace
-  view->horizontalHeader ()->restoreState (
-    settings->value ("workspaceview/column_state").toByteArray ());
-  // Make the header clickable to enable sorting
-  view->horizontalHeader ()->setClickable (true);
+  view->horizontalHeader ()->setSortIndicatorShown (true);
 
   // Init state of the filter
   _filter->addItems (settings->value ("workspaceview/mru_list").toStringList ());
 
   bool filter_state =
     settings->value ("workspaceview/filter_active", false).toBool ();
   _filter_checkbox->setChecked (filter_state);
   filter_activate (filter_state);
@@ -155,16 +157,24 @@ workspace_view::~workspace_view (void)
 
 void workspace_view::setModel (workspace_model *model)
 {
   _filter_model.setSourceModel (model);
   _filter_model.setFilterKeyColumn(0);
 
   view->setModel (&_filter_model);
 
+  // set the sorting after a model was set, it would be ignored otherwise
+  QSettings *settings = resource_manager::get_settings ();
+  view->sortByColumn (
+    settings->value ("workspaceview/sort_by_column",0).toInt (),
+    static_cast<Qt::SortOrder>
+    (settings->value ("workspaceview/sort_order", Qt::AscendingOrder).toUInt ())
+  );
+
   _model = model;
 }
 
 void
 workspace_view::closeEvent (QCloseEvent *e)
 {
   emit active_changed (false);
   QDockWidget::closeEvent (e);
